openedx/core/djangoapps/safe_endpoints/tests/test_middleware.py

```diff
@@ -0,0 +1,77 @@
+"""
+Unit tests for safe_endpoints Middleware.
+"""
+import ddt
+
+from django.test import TestCase
+from rest_framework.authentication import SessionAuthentication
+from rest_framework_jwt.authentication import JSONWebTokenAuthentication
+from rest_framework.decorators import api_view
+from rest_framework.views import APIView
+
+from openedx.core.djangolib.testing.utils import get_mock_request
+
+from ..middleware import EnsureJWTAuthSettingsMiddleware
+
+
+class MockIncludedPermissionClass(object): pass
+class MockRequiredPermissionClass(object): pass
+
+
+def mock_auth_decorator(include_jwt_auth=True):
+    def _decorator(f):
+        f.permission_classes = (MockIncludedPermissionClass,)
+        f.authentication_classes = (SessionAuthentication,)
+        if include_jwt_auth:
+            f.authentication_classes += (JSONWebTokenAuthentication,)
+        return f
+    return _decorator
+
+
+@api_view(["GET"])
+@mock_auth_decorator(include_jwt_auth=False)
+def mock_function_view(request): pass
+
+
+@api_view(["GET"])
+@mock_auth_decorator(include_jwt_auth=True)
+def mock_function_view_with_jwt(request): pass
+
+
+@mock_auth_decorator(include_jwt_auth=False)
+class MockClassAPIView(APIView):
+    pass
+
+
+@mock_auth_decorator(include_jwt_auth=True)
+class MockClassAPIViewWithJwt(APIView):
+    pass
+
+
+@ddt.ddt
+class TestEnsureJWTAuthSettingsMiddleware(TestCase):
+    def setUp(self):
+        super(TestEnsureJWTAuthSettingsMiddleware, self).setUp()
+        self.request = get_mock_request()
+
+    @ddt.data(
+        (mock_function_view, False),
+        (mock_function_view_with_jwt, True),
+        (MockClassAPIView, False),
+        (MockClassAPIViewWithJwt, True),
+    )
+    @ddt.unpack
+    def test_views(self, view, includes_required_permission):
+        middleware = EnsureJWTAuthSettingsMiddleware()
+        middleware.required_permission_classes = (MockRequiredPermissionClass,)
```

Comment:
Would it be a little safer to not mock here?  You could create a temporary fake class named `MockJwtHasScope` that is added to the real `required_permission_classes` and is not mocked in the test code.  The test code could temporarily look for `MockJwtHasScope`, and later would look for `JwtHasScope`.

Since this isn't meant to be reusable code, the real thing we want to ensure is that `JwtHasScope` is included.  In theory, if someone deleted the real line of code that put `JwtHasScope` in the `required_permission_classes`, we wouldn't have a test that failed if this just was based on mocks.

Reply:
I'm not really mocking here - as much as stubbing with another permission class for now.  Once I integrate with `JwtHasScope` I can remove this line of code and update the tests to make sure `JwtHasScope` gets included.

Is that what you're suggesting?

Reply:
Yes.  Got it.